src/freenet/client/async/KeyListenerTracker.java

```diff
@@ -95,61 +102,197 @@ protected static int fixRetryCount(int retryCount) {
 		return Math.max(0, retryCount-MIN_RETRY_COUNT);
 	}
 
+	private boolean contains(KeyListener[] listeners, KeyListener listener) {
+		for(KeyListener l : listeners) {
+			if(l == listener) return true;
```

Comment:
`.equals()` should be used here instead of comparing `KeyListener`s by identity.